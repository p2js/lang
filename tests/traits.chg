trait Default {
    fn default(): Self;
    fn method(self);
}
impl Default for i32 {
    fn default(): Self {
        return 0;
    }
    fn method(self) {}
}

fn main(): i32 {
    let x: i32 = 10;
    x.method();
    return Default::default();
}
