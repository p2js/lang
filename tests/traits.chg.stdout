expr @ 7:16 7:17 => i32
expr @ 7:9 7:17 => never
expr @ 6:24 9:0 => ()
expr @ 9:21 10:0 => ()
expr @ 13:18 13:20 => i32
expr @ 14:5 14:6 => i32
expr @ 14:5 14:15 => ()
expr @ 15:21 15:28 => fn default(...)
expr @ 15:21 15:30 => i32
expr @ 15:5 15:30 => never
expr @ 12:16 17:1 => ()



define void @method$LT$i32$GT$(i32 %lt.0) {
prologue:
%l.0 = alloca i32
store i32 %lt.0, i32* %l.0
br label %bb.0

bb.0:
ret void

}

define i32 @default$LT$i32$GT$() {
prologue:
%l.0 = alloca i32
br label %bb.0

bb.0:
store i32 0, i32* %l.0
%ret.0 = load i32, i32* %l.0
ret i32 %ret.0

}

define i32 @main() {
prologue:
%l.0 = alloca i32
%l.1 = alloca i32
br label %bb.0

bb.0:
store i32 10, i32* %l.1
%l.2 = load i32, i32* %l.1
%l.3 = call {} bitcast(void(i32)* @method$LT$i32$GT$ to {}(i32)*)(i32 %l.2)
br label %bb.1

bb.1:
%l.4 = call i32 @default$LT$i32$GT$()
br label %bb.2

bb.2:
store i32 %l.4, i32* %l.0
%ret.0 = load i32, i32* %l.0
ret i32 %ret.0

}


